Dark storage performance

mortalisk opened the pull request:
**Issue**
Resolves #2651 


**Approach**
Adds a template for instantiation of a forward model of required size. This is then run and used as a basis for testing the performance of various dark_storage functions.

There is a job on jenkins, https://ci.equinor.com/komodo/job/ert-performance/
This job runs with large and slow data sizes, by passing the flag --runslow to pytest
The test also runs with a small data size while running integration tests on github actions

Was originally worked on in #2704, but the test added there would not pass, so have created another version here, which does not pass through the TestClient, but rather calls the enpoint definitions directly.

mortalisk commented:
Jenkins test this please

oysteoh commented:
should pytest-benchmark be added as a requirement ( or dev-requirement? )

oysteoh reviewed (approved):
Refactoring removing the global state seems like a good choice - very nice 🚀

DanSava commented on ert_shared/dark_storage/enkf.py:
```diff
@@ -46,17 +34,19 @@ def get_name(type, uuid):
-    if _libres_facade is None:
-        init_facade()
-    return _libres_facade
+    configfile = os.environ["ERT_STORAGE_RES_CONFIG"]
+    config = ResConfig(configfile)
+    os.chdir(config.config_path)
+    ert = EnKFMain(config, strict=True)
+    notifier = ErtNotifier(ert, config)
+    adapter = ErtAdapter()
+    with adapter.adapt(notifier):
+        yield adapter.enkf_facade
```
Does this mean that we are going to initialize a new instance of EnkMain every time a new request is handled by dark_storage?

oysteoh replied:
i haven't thought about this, removing my approval until we have clarified

mortalisk replied:
Undid these changes

DanSava commented on ert_shared/dark_storage/common.py:
```diff
@@ -7,27 +9,29 @@
 
 
-def ensemble_parameters(ensemble_name: str) -> List[dict]:
+def ensemble_parameters(res, ensemble_name: str) -> List[dict]:
```
Type annotation missing for `res` argument

DanSava commented on ert_shared/dark_storage/common.py:
```diff
@@ -7,27 +9,29 @@
-def get_responses(ensemble_name: str):
-    res = get_res()
-    response_names = get_response_names()
+def get_responses(res, ensemble_name: str):
```
Type annotation missing for `res` argument

DanSava commented on ert_shared/dark_storage/common.py:
```diff
@@ -37,11 +41,10 @@ def get_responses(ensemble_name: str):
 
 
-def data_for_key(case, key, realization_index=None):
+def data_for_key(res, case, key, realization_index=None):
```
Type annotation missing for `res` argument

DanSava commented on ert_shared/dark_storage/common.py:
```diff
@@ -61,12 +64,11 @@ def data_for_key(case, key, realization_index=None):
 
 
-def observations_for_obs_keys(case, obs_keys):
+def observations_for_obs_keys(res, case, obs_keys):
```
Type annotation missing for `res` argument

DanSava commented on tests/ert_tests/dark_storage/test_api_compatibility.py:
```diff
@@ -12,6 +12,8 @@ def test_openapi(ert_storage_app, dark_storage_app):
     expect = ert_storage_app.openapi()
     actual = dark_storage_app.openapi()
 
+    # Remove /gql endpoint, as this is a workaround to get EnkfMain injected in the request object
```
I don't really understand this comment

mortalisk replied:
:brain:

DanSava commented on test-data/local/poly_template/poly_eval.py.j2:
```diff
@@ -0,0 +1,100 @@
+    return coeffs["c0"] * x ** 2 + coeffs["c1"] * x + coeffs["c2"]
+
+
+def f(x):
```
Very undescriptive function name

DanSava commented on test-data/local/poly_template/poly_eval.py.j2:
```diff
@@ -0,0 +1,100 @@
+
+
+def f(x):
+    return '{0: <8}'.format(x)
```
Why not use f"{x: <8}" to avoid "####".format() syntax

DanSava commented on tests/performance_tests/conftest.py:
```diff
@@ -0,0 +1,97 @@
+
+    yield params
+
+    # shutil.rmtree(poly_folder)
```
Leftover comment?

DanSava commented on tests/performance_tests/conftest.py:
```diff
@@ -0,0 +1,97 @@
+            ],
+        )
+        run_cli(parsed)
+        print(poly_folder)
```
Print statement, is it really needed?

DanSava commented on tests/performance_tests/conftest.py:
```diff
@@ -0,0 +1,97 @@
+        )
+        run_cli(parsed)
+        print(poly_folder)
+        files = list(x for x in Path(".").rglob("*") if "storage" not in x.parts)
+        files.sort()
+        for file in files:
+            print(f"{file}, {os.stat(file).st_size}")
```
This just prints a bunch of files and their size. Is that really required?

DanSava commented on tests/performance_tests/performance_utils.py:
```diff
@@ -0,0 +1,176 @@
+import sys
+
+
+def f(x):
```
Very undescriptive function name

DanSava commented on ert_shared/dark_storage/enkf.py:
```diff
@@ -46,17 +34,19 @@ def get_name(type, uuid):
 
-def get_size(ensemble_name):
-    res = get_res()
+def get_size(res, ensemble_name):
```
Does not really use the ensemble_name

DanSava commented on ert_shared/dark_storage/endpoints/experiments.py:
```diff
@@ -58,7 +58,7 @@ def get_experiment_ensembles(
             experiment_id=get_id("experiment", "default"),
             userdata={"name": case},
-            size=get_size(case),
+            size=get_size(res, case),
```
`get_size` does nothing with the case name, maybe we should remove it

DanSava commented on ert_shared/dark_storage/endpoints/ensembles.py:
```diff
@@ -32,9 +32,11 @@ def get_ensemble(
-        size=get_size(get_name("ensemble", ensemble_id)),
-        parameter_names=ensemble_parameter_names(get_name("ensemble", ensemble_id)),
-        response_names=get_response_names(),
+        size=get_size(res, get_name("ensemble", ensemble_id)),
```
`get_size` does nothing with the case name, maybe we should remove it

oysteoh commented:
Another question popped up while discussing this - how is the jenkins job thought to be ran, and how long time will it take?

mortalisk commented on ert_shared/async_utils.py:
```diff
@@ -7,3 +7,7 @@ def get_event_loop() -> asyncio.AbstractEventLoop:
         return asyncio.get_event_loop()
+
+
+def run_in_loop(coro):
```
I don't know what type that is supposed to be, do you?

mortalisk replied:
I added some typing. There is a chance it is correct. Would mypy complain if it was incorrect?

mortalisk commented on ci/jenkins/Jenkinsfile-performance:
```diff
@@ -0,0 +1,43 @@
+                pytest tests/performance_tests  --benchmark-autosave --benchmark-storage=file://${env.BASE_DIR}/.benchmarks --basetemp=${env.TMP_DIR} ${PYTEST_OPTIONS}
+                """
+            }
+            post {
+                failure {
+                    echo "failure"
+                }
+                success {
+                    echo "success"
+                }
+                cleanup {
+                    echo "cleanup"
+                }
```
No, just left over

mortalisk commented:
I have changed the code a bit to address the issues mentioned. Most importantly, I have removed the changes to the initialization of enkf_main in dark_storage, but kept the ability to manually inject it in tests, which is the key point my test needs to function at the moment. This was done by changing the first commit. The other changes should be in additional commits. I will reorder and squash commits in some way that makes sense before merging.

I have also added some asserts in get_observations, other than that, the changes are mostly cosmetic, and according to review comments.

mortalisk commented on tests/ert_tests/dark_storage/conftest.py:
```diff
@@ -57,7 +57,8 @@ def poly_example_tmp_dir(poly_example_tmp_dir_shared):
         monkeypatch.setenv("ERT_STORAGE_RES_CONFIG", "poly.ert")
-        yield TestClient(dark_app)
+        with TestClient(dark_app) as client:
+            yield client
```
I think this is the correct way of doing this. Otherwise, fastapi shutdown code won't run. We don't need that now, but probably best to do it right, in case we need it at some point.

mortalisk commented on tests/ert_tests/dark_storage/test_api_compatibility.py:
```diff
@@ -12,6 +12,10 @@ def test_openapi(ert_storage_app, dark_storage_app):
     actual = dark_storage_app.openapi()
 
+    # This endpoint is added explicitly only in dark_storage, to have a place to inject
+    # LibresFacade. Remove it here to make it the same as ert_storage.
```
Hope this comment is understandable

mortalisk commented on tests/performance_tests/performance_utils.py:
```diff
@@ -0,0 +1,170 @@
+    if not os.path.exists(folder / "refcase"):
+        os.mkdir(folder / "refcase")
+
+    use_ecl_data_io = True
```
Should we keep the ability to switch to using EclSum for creating summary data?

oysteoh commented:
I think before this PR is merged, there should be some sort of baseline stored along with the test providing us with the ability of comparison of runs. Only running jenkins nightly and require us to poke into the results will not make this testing very useful. If we could save and establish a baseline now documenting the performance of the api, we have something to always check against and flag / report error the moment we are introducing some code degrading the performance.

mortalisk commented:
> 

On the jenkins job there should be some files stored from the previous run. I am unsure if it is used right now for anything at all, but it should be possible to make pytest-benchmark compare and fail if there is degradation. Could we perhaps make another issue to do more work on that Jenkins job? I think it's a good idea to get this functionality in main.

oysteoh commented:
yup, we could do that. But - i think the baseline should be stored in the repo itself providing a fixed set of numbers to always compare against up until we decide to adjust? If we always compare with the previous one we could potentially experience a degradation over time without noticing it?

oysteoh commented on ert_shared/dark_storage/common.py:
```diff
@@ -1,33 +1,34 @@
-def get_responses(ensemble_name: str):
-    res = get_res()
-    response_names = get_response_names()
+def get_responses(res: LibresFacade, ensemble_name: str):
```
missing return type

oysteoh commented on ert_shared/dark_storage/common.py:
```diff
@@ -37,11 +38,10 @@ def get_responses(ensemble_name: str):
 
 
-def data_for_key(case, key, realization_index=None):
+def data_for_key(res: LibresFacade, case, key, realization_index=None):
```
missing return type

oysteoh commented:
This now seems ok from my point of view. Could you fix the commit-history of this PR as you wanted before i approve? And - also make an issue regarding creating some sort of baseline which this performance-test could use when its running?

mortalisk commented:
Made #3280 for the baseline work

oysteoh reviewed (approved):
LGTM!

oysteoh commented:
Jenkins: Test this please!

oysteoh commented:
Jenkins: Test this please!
